### PR TITLE
[feat] cuequivalence triangle multilplicative update

### LINKED ISSUE #162: Dead branch?
https://github.com/bytedance/Protenix/blob/128b33bede180df9737f8be245f7ba5e08c03db1/protenix/openfold_local/model/triangular_multiplicative_update.py#L391

Is this code ever executed? Seems like there's a missing reshape (at the least?) after the matmul...

### PR DESCRIPTION
**Pull request by popfido**
## What does this PR do?

This PR should be merged after #164 . Introducing optimized fusion kernel for triangle multilplicative update from `cuequivalence`.

Also fixed #162 

Add triangle model components and optimizations
- Introduced new files for triangle model components including `fused_layer_norm`, `gated_gemm`, and `triangle_multiplicative_update`.
- Added `fused_sigmoid_gated_dual_gemm` functions for optimized matrix multiplication with gating mechanisms.
- Enhanced the `triangle_multiplicative_update` function to support various input configurations and normalization.
- Included tests for the triangular multiplicative update to ensure functionality and performance improvements.

These changes aim to improve the efficiency and maintainability of the triangle model operations in the Protenix framework.

Why separately extract implementation from `cuequivariance`:

- Installation & Importing of `cuequivariance` including 2 packages and relates to different implementation (`jax` / `pytorch`). Adding the complexity of maintaining flexible environment.
- Core logic of fusion triangle_multiplicative_update is enough separated as a single module to be imported.
- `cuequivariance` is of Apache License v2.

Tests:

```
python -m unittest tests.test_triangular_multiplicative_update.TestTriangularMultiplicativeUpdateMemory
```

TriangularMultiplicativeUpdate Forward - Time vs N (D=128, Outgoing, Linear Scale)

TriangularMultiplicativeUpdate Forward - Peak Memory vs N (D=128, Outgoing, Linear Scale)

Percentage Time Reduction vs Normal (D=128, Outgoing)

Percentage Peak Memory Reduction vs Normal (D=128, Outgoing)

Backward - Time vs N (D=128, Linear Scale)

Backward - Peak Memory vs N (D=128, Linear Scale)

Backward - Time Reduction vs Standard (D=128)

Backward - Peak Memory Reduction vs Standard (D=128)

All tests ran under a single A800 - 80G card with PyTorch 2.5.1 and Triton 3.1 for reference. (Since my PTX env is not quite standard as I upgraded some of the dependency version)

Noted that `cuequivalence` has been reported to have some [backward issue](https://github.com/NVIDIA/cuEquivariance/issues/144) when facing super long sequence (>2800). But it should not affect this implementation since for training / finetuning, the input length won't be so long. While for long sequence inference, there is no need for taking care of backward issue. And Chunking + Inplace fusion Implementation works well till 4096 according to my test. Thus current implementation is quite stable.

All Licence from NVIDIA has been preserved. While it is strange that `cuequivalence` itself has  Apache License v2, all its internal files inside the project has SPDX-License, which is inconsistent. I just treat it as a typo issue and will update the licence to `Apache License v2` when NVIDIA has confirmed.

I'll leave adding of a switcher for this fusion kernel in config file to project owner to decide. If it needs my help, please assign to me.

## Break Changes

- None

## Before submitting

- [x] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **run pre-commit hooks** with `pre-commit run -a` command?

### CODE REVIEW

**Comment by zhangyuxuann** [comment]
@popfido  Thank you for your contribution! @yangyanpinghpc  will review the PR.

**Comment by yangyanpinghpc** [comment]
@popfido  Before we review the MR code, I have a few questions.
1. Could you explain why the API cuequivalence_triangle_multiplicative_update isn't used directly?
2. The license information in your code indicates that NVIDIA prohibits copying, moving, or reproducing this code. and I couldn't find the files fused_layer_norm.py and gated_gemm.py in the NVIDIA repo.  Could you please provide the URL for the code?

**Comment by mrzzmrzz** [comment]
Since the issue in cuEquivariance [triangle_multiplicative_update crash when length over 2816 #144](https://github.com/NVIDIA/cuEquivariance/issues/144) has not yet been resolved, directly using the built-in `cuequivalence_triangle_multiplicative_update` may cause illegal memory access errors during inference for sequences that are too long. Therefore, I think it might not be advisable to use it directly until this bug is fixed.

In addition, I noticed that `fused_layer_norm.py` and `gated_gemm.py` correspond to `fused_layer_norm_torch.py` and `gated_gemm_torch.py` in `cuequivariance_ops_torch` package, respectively.

I hope the above information is helpful to you.

**Comment by rakeshr10** [comment]
> Since the issue in cuEquivariance [triangle_multiplicative_update crash when length over 2816 #144](https://github.com/NVIDIA/cuEquivariance/issues/144) has not yet been resolved, directly using the built-in `cuequivalence_triangle_multiplicative_update` may cause illegal memory access errors during inference for sequences that are too long. Therefore, I think it might not be advisable to use it directly until this bug is fixed.
> 
> In addition, I noticed that `fused_layer_norm.py` and `gated_gemm.py` correspond to `fused_layer_norm_torch.py` and `gated_gemm_torch.py` in `cuequivariance_ops_torch` package, respectively.
> 
> I hope the above information is helpful to you.

This is resolved in the recent [v0.6.0](https://github.com/NVIDIA/cuEquivariance/releases/tag/v0.6.0) release

**Comment by mrzzmrzz** [comment]
> > Since the issue in cuEquivariance [triangle_multiplicative_update crash when length over 2816 #144](https://github.com/NVIDIA/cuEquivariance/issues/144) has not yet been resolved, directly using the built-in `cuequivalence_triangle_multiplicative_update` may cause illegal memory access errors during inference for sequences that are too long. Therefore, I think it might not be advisable to use it directly until this bug is fixed.
> > In addition, I noticed that `fused_layer_norm.py` and `gated_gemm.py` correspond to `fused_layer_norm_torch.py` and `gated_gemm_torch.py` in `cuequivariance_ops_torch` package, respectively.
> > I hope the above information is helpful to you.
> 
> This is resolved in the recent [v0.6.0](https://github.com/NVIDIA/cuEquivariance/releases/tag/v0.6.0) release

I tested it, and in the latest version 0.6.0, when the token length exceeds 2880, it still triggers an illegal memory access error (Triton Error [CUDA]: an illegal memory access was encountered).

**Comment by rakeshr10** [comment]
> > > > Since the issue in cuEquivariance [triangle_multiplicative_update crash when length over 2816 #144](https://github.com/NVIDIA/cuEquivariance/issues/144) has not yet been resolved, directly using the built-in `cuequivalence_triangle_multiplicative_update` may cause illegal memory access errors during inference for sequences that are too long. Therefore, I think it might not be advisable to use it directly until this bug is fixed.
> > > > In addition, I noticed that `fused_layer_norm.py` and `gated_gemm.py` correspond to `fused_layer_norm_torch.py` and `gated_gemm_torch.py` in `cuequivariance_ops_torch` package, respectively.
> > > > I hope the above information is helpful to you.
> > > 
> > > 
> > > This is resolved in the recent [v0.6.0](https://github.com/NVIDIA/cuEquivariance/releases/tag/v0.6.0) release
> > 
> > 
> > I tested it, and in the latest version 0.6.0, when the token length exceeds 2880, it still triggers an illegal memory access error (Triton Error [CUDA]: an illegal memory access was encountered).
> 
> I have also tested it and confirmed that version 0.6.0 does not fix this issue.After reviewing the triton code, I found that the problem is caused by data flow.
> 
> cuequivariance_ops/triton/gated_gemm_triton.py if TRANSPOSE_OUT: o_ptrs = o_ptr + tl.cast(offs_on[None, :], tl.int64) * tl.cast(M, tl.int64) + tl.cast(offs_om[:, None], tl.int64) else: o_ptrs = o_ptr + tl.cast(offs_om[:, None], tl.int64) * tl.cast(N, tl.int64) + tl.cast(offs_on[None, :], tl.int64)

Thanks for the information. Is there a way you can report this issue to the Nvidia team on the library github site?

**Comment by yangyanpinghpc** [comment]
> > > > > Since the issue in cuEquivariance [triangle_multiplicative_update crash when length over 2816 #144](https://github.com/NVIDIA/cuEquivariance/issues/144) has not yet been resolved, directly using the built-in `cuequivalence_triangle_multiplicative_update` may cause illegal memory access errors during inference for sequences that are too long. Therefore, I think it might not be advisable to use it directly until this bug is fixed.
> > > > > In addition, I noticed that `fused_layer_norm.py` and `gated_gemm.py` correspond to `fused_layer_norm_torch.py` and `gated_gemm_torch.py` in `cuequivariance_ops_torch` package, respectively.
> > > > > I hope the above information is helpful to you.
> > > > 
> > > > 
> > > > This is resolved in the recent [v0.6.0](https://github.com/NVIDIA/cuEquivariance/releases/tag/v0.6.0) release
> > > 
> > > 
> > > I tested it, and in the latest version 0.6.0, when the token length exceeds 2880, it still triggers an illegal memory access error (Triton Error [CUDA]: an illegal memory access was encountered).
> > 
> > 
> > I have also tested it and confirmed that version 0.6.0 does not fix this issue.After reviewing the triton code, I found that the problem is caused by data flow.
> > cuequivariance_ops/triton/gated_gemm_triton.py if TRANSPOSE_OUT: o_ptrs = o_ptr + tl.cast(offs_on[None, :], tl.int64) * tl.cast(M, tl.int64) + tl.cast(offs_om[:, None], tl.int64) else: o_ptrs = o_ptr + tl.cast(offs_om[:, None], tl.int64) * tl.cast(N, tl.int64) + tl.cast(offs_on[None, :], tl.int64)
> 
> Thanks for the information. Is there a way you can report this issue to the Nvidia team on the library github site?

Since the cuequivariance Triton codes aren't included in the cuequivariance repository, we're unable to file a merge request. But, we will submit an issue in the cuequivariance repository and attempt to make contact with them.

**Comment by yangyanpinghpc** [comment]
We have integrated cuEquivariance into our Protenix repository. Due to license restrictions, we are unable to include the source code of cuEquivariance. We have filed an issue with NVIDIA regarding this matter (titled "Index Overflow & Need for Chunk Operation Support," accessible at https://github.com/NVIDIA/cuEquivariance/issues/154) and are currently awaiting their response.

**Comment by zhangyuxuann** [comment]
Thank you so much for your contribution — we really appreciate the time and effort you've put into this!

Unfortunately, after careful review, we're unable to merge this PR due to licensing restrictions. The code includes dependencies or components that are not compatible with our project's open-source license, and we must ensure full compliance.

We’ll close this PR for now.

Thanks again for your understanding and support!

**Comment by zhangyuxuann** [comment]
Thank you so much for your contribution — we really appreciate the time and effort you've put into this!

Unfortunately, after careful review, we're unable to merge this PR due to licensing restrictions. The code includes dependencies or components that are not compatible with our project's open-source license, and we must ensure full compliance.

We’ll close this PR for now.

Thanks again for your understanding and support!

**Comment by kucukben** [comment]
Hi from cuEquivariance team! Thanks for your interest in our library & for reporting the bug. We are preparing a patch release for this. Will keep you updated on the issue you opened. Just a quick comment on the license: frontend is licensed under Apache v.2, backend (where you found these triton files) are NVIDIA proprietary licensed, so it is not a typo. Thanks again.

**Comment by hsadasivan** [comment]
cueq 0.6.1 is now released with the discussed fix.
